File: kotlin/src/commonTest/kotlin/com/actualbudget/sync/crdt/TimestampTest.kt
package com.actualbudget.sync.crdt

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class TimestampTest {

    @Test
    fun testParseValidTimestamp() {
        val ts = Timestamp.parse("2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912")
        assertNotNull(ts)
        assertEquals(1745533422123, ts.millis)
        assertEquals(1, ts.counter)
        assertEquals("A219E7A71CC18912", ts.node)
    }

    @Test
    fun testParseInvalidTimestamp() {
        assertNull(Timestamp.parse("invalid"))
        assertNull(Timestamp.parse("2025-04-24T22:23:42.123Z"))
        assertNull(Timestamp.parse("2025-04-24T22:23:42.123Z-0001"))
    }

    @Test
    fun testTimestampToString() {
        val ts = Timestamp(1745533422123, 1, "A219E7A71CC18912")
        val str = ts.toString()
        assertTrue(str.contains("2025-04-24"))
        assertTrue(str.contains("-0001-"))
        assertTrue(str.contains("A219E7A71CC18912"))
    }

    @Test
    fun testTimestampRoundTrip() {
        val original = "2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912"
        val ts = Timestamp.parse(original)
        assertNotNull(ts)
        val result = ts.toString()
        // Parse and re-serialize
        val ts2 = Timestamp.parse(result)
        assertNotNull(ts2)
        assertEquals(ts.millis, ts2.millis)
        assertEquals(ts.counter, ts2.counter)
        assertEquals(ts.node, ts2.node)
    }

    @Test
    fun testMakeClientId() {
        val id1 = Timestamp.makeClientId()
        val id2 = Timestamp.makeClientId()
        assertEquals(16, id1.length)
        assertEquals(16, id2.length)
        assertTrue(id1 != id2, "Client IDs should be unique")
    }

    @Test
    fun testTimestampHash() {
        val ts1 = Timestamp.parse("2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912")!!
        val ts2 = Timestamp.parse("2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912")!!
        val ts3 = Timestamp.parse("2025-04-24T22:23:42.123Z-0002-A219E7A71CC18912")!!

        assertEquals(ts1.hash(), ts2.hash())
        assertTrue(ts1.hash() != ts3.hash())
    }

    @Test
    fun testTimestampComparison() {
        val ts1 = Timestamp.parse("2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912")!!
        val ts2 = Timestamp.parse("2025-04-24T22:23:42.123Z-0002-A219E7A71CC18912")!!
        val ts3 = Timestamp.parse("2025-04-25T22:23:42.123Z-0001-A219E7A71CC18912")!!

        assertTrue(ts1 < ts2)
        assertTrue(ts2 < ts3)
        assertTrue(ts1 < ts3)
    }

    @Test
    fun testZeroTimestamp() {
        assertEquals(0, Timestamp.ZERO.millis)
        assertEquals(0, Timestamp.ZERO.counter)
        assertEquals("0000000000000000", Timestamp.ZERO.node)
    }
}

File: dart/test/crdt/timestamp_test.dart
import 'package:test/test.dart';
import 'package:actual_sync/src/crdt/timestamp.dart';

void main() {
  group('Timestamp', () {
    test('parse valid timestamp', () {
      final ts = Timestamp.parse('2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912');
      expect(ts, isNotNull);
      expect(ts!.millis, equals(1745533422123));
      expect(ts.counter, equals(1));
      expect(ts.node, equals('A219E7A71CC18912'));
    });

    test('parse invalid timestamp', () {
      expect(Timestamp.parse('invalid'), isNull);
      expect(Timestamp.parse('2025-04-24T22:23:42.123Z'), isNull);
      expect(Timestamp.parse('2025-04-24T22:23:42.123Z-0001'), isNull);
    });

    test('toString', () {
      final ts = Timestamp(1745533422123, 1, 'A219E7A71CC18912');
      final str = ts.toString();
      expect(str, contains('2025-04-24'));
      expect(str, contains('-0001-'));
      expect(str, contains('A219E7A71CC18912'));
    });

    test('round trip', () {
      const original = '2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912';
      final ts = Timestamp.parse(original);
      expect(ts, isNotNull);
      final result = ts!.toString();
      // Parse and re-serialize
      final ts2 = Timestamp.parse(result);
      expect(ts2, isNotNull);
      expect(ts2!.millis, equals(ts.millis));
      expect(ts2.counter, equals(ts.counter));
      expect(ts2.node, equals(ts.node));
    });

    test('makeClientId', () {
      final id1 = Timestamp.makeClientId();
      final id2 = Timestamp.makeClientId();
      expect(id1.length, equals(16));
      expect(id2.length, equals(16));
      expect(id1, isNot(equals(id2)));
    });

    test('hash', () {
      final ts1 = Timestamp.parse('2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912')!;
      final ts2 = Timestamp.parse('2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912')!;
      final ts3 = Timestamp.parse('2025-04-24T22:23:42.123Z-0002-A219E7A71CC18912')!;

      expect(ts1.hash(), equals(ts2.hash()));
      expect(ts1.hash(), isNot(equals(ts3.hash())));
    });

    test('comparison', () {
      final ts1 = Timestamp.parse('2025-04-24T22:23:42.123Z-0001-A219E7A71CC18912')!;
      final ts2 = Timestamp.parse('2025-04-24T22:23:42.123Z-0002-A219E7A71CC18912')!;
      final ts3 = Timestamp.parse('2025-04-25T22:23:42.123Z-0001-A219E7A71CC18912')!;

      expect(ts1.compareTo(ts2), lessThan(0));
      expect(ts2.compareTo(ts3), lessThan(0));
      expect(ts1.compareTo(ts3), lessThan(0));
    });

    test('zero timestamp', () {
      expect(Timestamp.zero.millis, equals(0));
      expect(Timestamp.zero.counter, equals(0));
      expect(Timestamp.zero.node, equals('0000000000000000'));
    });
  });

  group('MutableClock', () {
    test('send increments counter', () {
      final clock = MutableClock(millis: 1700000000000, counter: 0, node: 'TESTNODEID123456');

      // Send multiple times quickly - counter should increment
      final ts1 = clock.send();
      final ts2 = clock.send();
      final ts3 = clock.send();

      // If all in same millisecond, counters should increment
      if (ts1.millis == ts2.millis && ts2.millis == ts3.millis) {
        expect(ts1.counter, lessThan(ts2.counter));
        expect(ts2.counter, lessThan(ts3.counter));
      }
    });

    test('recv merges timestamps', () {
      final clock = MutableClock(node: 'LOCALNODEID12345');
      final remoteTs = Timestamp(
        1700000001000, // Fixed future timestamp
        5,
        'REMOTENODEID1234',
      );

      final result = clock.recv(remoteTs);
      expect(result, isNotNull);

      // Clock should have advanced
      expect(clock.millis, greaterThanOrEqualTo(remoteTs.millis));
    });
  });
}
